src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java

```diff
@@ -100,6 +162,14 @@ public boolean processRow(StepMetaInterface smi, StepDataInterface sdi) throws K
 
                 setOutputDone();
 
+                //now output rowset closed, so we can wait for our confirmation steps
+                if ( data.activeConfirmation )  {
+                    while (!isStopped() && data.watchedConfirmStep.size() > 0) {    
+                        Thread.sleep(500);

```

Comment:
The ``transFinished`` gets called at the end of the transformation
with means all messages already went through the ack/nack step
and all tags have been collected be the RowListener.


The current behaviour call flush on ``dispose`` if transactional is not enabled, 
so you would have to also check for ``data.activeConfirmation`` before call ``flush`` and then delay it to ``transFinished``